File: CS/E4589/ViewModel.cs
using System.ComponentModel;
using System.Globalization;
using System.Windows;
using System.Windows.Input;
using DevExpress.Xpf.Core.Commands;
using DevExpress.Xpf.Printing;

namespace E4589 {
    class ViewModel : INotifyPropertyChanged {
        string[] data;
        DelegateCommand<object> createDocumentCommand;
        DelegateCommand<object> clearDocumentCommand;
        LinkPreviewModel previewModel;

        public IDocumentPreviewModel PreviewModel {
            get {
                if(previewModel == null) {
                    previewModel = new LinkPreviewModel();
                    previewModel.Link = CreateLink();
                }
                return previewModel;

            }
        }
        public ICommand CreateDocumentCommand {
            get {
                if(createDocumentCommand == null)
                    createDocumentCommand = new DelegateCommand<object>(ExecuteCreateDocumentCommand, CanExecuteCreateDocumentCommand);
                return createDocumentCommand;
            }
        }
        public ICommand ClearDocumentCommand {
            get {
                if(clearDocumentCommand == null)
                    clearDocumentCommand = new DelegateCommand<object>(ExecuteClearDocumentCommand, CanExecuteClearDocumentCommand);
                return clearDocumentCommand;
            }
        }

        bool CanExecuteCreateDocumentCommand(object parameter) {
            return true;
        }
        void ExecuteCreateDocumentCommand(object parameter) {
            ((LinkPreviewModel)PreviewModel).Link.CreateDocument(true);
            clearDocumentCommand.RaiseCanExecuteChanged();
        }

        bool CanExecuteClearDocumentCommand(object parameter) {
            return !PreviewModel.IsEmptyDocument;
        }
        void ExecuteClearDocumentCommand(object parameter) {
            previewModel.Link.PrintingSystem.ClearContent();
            clearDocumentCommand.RaiseCanExecuteChanged();
        }

        LinkBase CreateLink() {
            data = CultureInfo.CurrentCulture.DateTimeFormat.DayNames;
            var link = new SimpleLink();
            link.DetailTemplate = (DataTemplate)Resources.Instance["dayNameTemplate"];
            link.DetailCount = data.Length;
            link.CreateDetail += link_CreateDetail;
            return link;
        }

        void link_CreateDetail(object sender, CreateAreaEventArgs e) {
            e.Data = data[e.DetailIndex];
        }

        #region INotifyPropertyChanged

        public event PropertyChangedEventHandler PropertyChanged;

        void RaisePropertyChanged(string propertyName) {
            if(PropertyChanged != null) {
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
            }
        }
        #endregion
    }
}

File: VB/E4589/ViewModel.vb
Imports System.ComponentModel
Imports System.Globalization
Imports System.Windows
Imports System.Windows.Input
Imports DevExpress.Xpf.Core.Commands
Imports DevExpress.Xpf.Printing

Namespace E4589
	Friend Class ViewModel
		Implements INotifyPropertyChanged

		Private data() As String
'INSTANT VB NOTE: The field createDocumentCommand was renamed since Visual Basic does not allow fields to have the same name as other class members:
		Private createDocumentCommand_Conflict As DelegateCommand(Of Object)
'INSTANT VB NOTE: The field clearDocumentCommand was renamed since Visual Basic does not allow fields to have the same name as other class members:
		Private clearDocumentCommand_Conflict As DelegateCommand(Of Object)
'INSTANT VB NOTE: The field previewModel was renamed since Visual Basic does not allow fields to have the same name as other class members:
		Private previewModel_Conflict As LinkPreviewModel

		Public ReadOnly Property PreviewModel() As IDocumentPreviewModel
			Get
				If previewModel_Conflict Is Nothing Then
					previewModel_Conflict = New LinkPreviewModel()
					previewModel_Conflict.Link = CreateLink()
				End If
				Return previewModel_Conflict

			End Get
		End Property
		Public ReadOnly Property CreateDocumentCommand() As ICommand
			Get
				If createDocumentCommand_Conflict Is Nothing Then
					createDocumentCommand_Conflict = New DelegateCommand(Of Object)(AddressOf ExecuteCreateDocumentCommand, AddressOf CanExecuteCreateDocumentCommand)
				End If
				Return createDocumentCommand_Conflict
			End Get
		End Property
		Public ReadOnly Property ClearDocumentCommand() As ICommand
			Get
				If clearDocumentCommand_Conflict Is Nothing Then
					clearDocumentCommand_Conflict = New DelegateCommand(Of Object)(AddressOf ExecuteClearDocumentCommand, AddressOf CanExecuteClearDocumentCommand)
				End If
				Return clearDocumentCommand_Conflict
			End Get
		End Property

		Private Function CanExecuteCreateDocumentCommand(ByVal parameter As Object) As Boolean
			Return True
		End Function
		Private Sub ExecuteCreateDocumentCommand(ByVal parameter As Object)
			DirectCast(PreviewModel, LinkPreviewModel).Link.CreateDocument(True)
			clearDocumentCommand_Conflict.RaiseCanExecuteChanged()
		End Sub

		Private Function CanExecuteClearDocumentCommand(ByVal parameter As Object) As Boolean
			Return Not PreviewModel.IsEmptyDocument
		End Function
		Private Sub ExecuteClearDocumentCommand(ByVal parameter As Object)
			previewModel_Conflict.Link.PrintingSystem.ClearContent()
			clearDocumentCommand_Conflict.RaiseCanExecuteChanged()
		End Sub

		Private Function CreateLink() As LinkBase
			data = CultureInfo.CurrentCulture.DateTimeFormat.DayNames
			Dim link = New SimpleLink()
			link.DetailTemplate = DirectCast(Resources.Instance("dayNameTemplate"), DataTemplate)
			link.DetailCount = data.Length
			AddHandler link.CreateDetail, AddressOf link_CreateDetail
			Return link
		End Function

		Private Sub link_CreateDetail(ByVal sender As Object, ByVal e As CreateAreaEventArgs)
			e.Data = data(e.DetailIndex)
		End Sub

		#Region "INotifyPropertyChanged"

		Public Event PropertyChanged As PropertyChangedEventHandler

		Private Sub RaisePropertyChanged(ByVal propertyName As String)
			RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
		End Sub
		#End Region
	End Class
End Namespace
